ee/tabby-db/src/google_oauth_credential.rs

```diff
@@ -0,0 +1,172 @@
+use anyhow::Result;
+use chrono::{DateTime, Utc};
+use rusqlite::{named_params, OptionalExtension};
+
+use super::DbConn;
+
+const GOOGLE_OAUTH_CREDENTIAL_ROW_ID: i32 = 1;
+
+pub struct GoogleOAuthCredentialDAO {
+    pub client_id: String,
+    pub client_secret: String,
+    pub redirect_uri: String,
+    pub created_at: DateTime<Utc>,
+    pub updated_at: DateTime<Utc>,
+}
+
+impl GoogleOAuthCredentialDAO {
+    fn from_row(row: &rusqlite::Row<'_>) -> std::result::Result<Self, rusqlite::Error> {
+        Ok(Self {
+            client_id: row.get(0)?,
+            client_secret: row.get(1)?,
+            redirect_uri: row.get(2)?,
+            created_at: row.get(3)?,
+            updated_at: row.get(4)?,
+        })
+    }
+}
+
+/// db read/write operations for `google_oauth_credential` table
+impl DbConn {
+    pub async fn update_google_oauth_credential(
+        &self,
+        client_id: &str,
+        client_secret: &str,
+        redirect_uri: Option<&str>,
+    ) -> Result<()> {
+        let redirect_uri = redirect_uri.unwrap_or_default().to_string();
+        if !client_secret.is_empty() {
```

Comment:
Not a big fan of this pattern, since it really balloons if we ever have more than one optional field. Maybe you could use a `format!` to construct the query, having it do `SET client_secret = :client_secret` when the secret is provided, and `SET client_secret = client_secret` for a no-op when it is not provided? That way just a tiny piece of data would be conditional, and the rest of the logic would have no branching and no additional queries.

Reply:
1. be aware that we never return `client_secret` to FE, so if passed secret is non-empty, very likely it's (re)set by the admin. So it goes to the `if` branch.
2. multi optional fields are fine, they all go to the `else` branch if the admin just wants to update them, e.g. `redirect_url` for google oauth
3. about your suggestion, I found there is one case that we might insert a record `(1, '', '', '')` because the schema requires `NOT NULL` for id, secret, redirect_url. So the insert statement is fixed as `insert into google_oauth_credential (id,client_id,client_secret,redirect_uri) values (...)`, we can only tune the `do update part`. But for current implementation, if secret is empty and no existing record, the `update` in else branch will fail as expected.

Reply:
Sounds good, then. Just wish there was a better way to do this.